### PR TITLE
[Deprecated][SQL] Add support for collation for StringTrim type of functions/expressions

### PR DESCRIPTION
**Pull request by davidm-db**
### Left to do/discussion topics (to be removed by the PR completion)
- Handling of a lowercase (`utf8_binary_lcase`) collation type seems a bit grueling in the long run. This PR handles it the same way as in every other place, since I think it is out of the scope of this PR, but should we maybe create a work item to generalize this collation type better (if we don't have it already)?

### What changes were proposed in this pull request?
This PR is created to add support for collations to StringTrim family of functions/expressions, specifically:
- `StringTrim`
- `StringTrimBoth`
- `StringTrimLeft`
- `StringTrimRight`

Changes:
- `stringExpressions.scala`
  - Check input types to be of the same collation.
  - Route to proper endpoints in `UTF8String` - `doEval` and `doGenCode`.
- `UTF8String.java`
  - Implement logic for all trim functions (logic should be well commented-out in the code and mostly matches already existing implementations of `trimLeft` and `trimRight`).

Other minor changes in this PR:
- Change parameter names in `CollationFactory.getStringSearch()` to a more descriptive ones.

### Why are the changes needed?
We are incrementally adding collation support to a built-in string functions in Spark.

### Does this PR introduce _any_ user-facing change?
Yes:
- User should now be able to use non-default collations in string trim functions.
- User will receive `DATATYPE_MISMATCH.COLLATION_MISMATCH` exception if collations of two function parameters do not match.

### How was this patch tested?
Already existing tests + new unit/e2e tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 1]
```diff
@@ -176,11 +176,11 @@ public Collation(
    */
 
   public static StringSearch getStringSearch(
-      final UTF8String left,
-      final UTF8String right,
+      final UTF8String targetString,
+      final UTF8String patternString,
```
Comment (mihailom-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:180]: Agreed that this should be changed, but I find it a bit weird. We call it `targetString` and then call `toString` method on it later. I would maybe add something like `targetUTF8String`. @dbatomic what do you think about this?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -733,27 +873,30 @@ public UTF8String trimRight(UTF8String trimString) {
...
           stringCharPos[numChars - 1],
           stringCharPos[numChars - 1] + stringCharLen[numChars - 1] - 1);
+
       if (trimString.find(searchChar, 0) >= 0) {
         trimEnd -= stringCharLen[numChars - 1];
       } else {
         break;
       }
-      numChars --;
+      numChars--;
```
Comment (mihailom-db) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:899]: Avoid reformatting the whole file, as it creates a lot of backporting issues. If you want to reformat your code select the part and pres cmd + option + L.

**Comment by mihailom-db** [comment]
Also, please add tests to `CollationStringExpressionSuite`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 2 above)
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:899]: Wasn't aware that this might be the issue, thanks!

**Comment by davidm-db** [comment]
@mihailom-db I'm adding tests to both `CollationStringExpressionSuite` and `CollationSuite` - I just pushed the changes so they can be reviewed while I'm adding the tests :)

**Comment by davidm-db** [comment]
@MaxGekk, @cloud-fan could someone please take a look at this PR?

Tagging the rest of Belgrade collation crew if anyone would like to review additionally: @dbatomic, @nikolamand-db, @stefankandic, @uros-db

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -686,6 +743,111 @@ public UTF8String trimLeft(UTF8String trimString) {
...
+    }
+
+    if (CollationFactory.UTF8_BINARY_LCASE_COLLATION_ID == collationId) {
+      return lowercaseTrimLeft(trimString);
+    }
+
+    return collatedTrimLeft(trimString, collationId);
+  }
+
+  private UTF8String lowercaseTrimLeft(UTF8String trimString) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:767]: Instead of implementing `lowercaseTrimLeft` and `collatedTrimLeft` separately (these functions look very similar to me), I think we could make use of new `StringSearch(pattern, target)` (with `.toLowerCase()` for both params, and no collationId for **UTF8_BINARY_LCASE**)

For more context, please take a look at: https://github.com/apache/spark/pull/45704/files#r1538624688

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 4]
```diff
@@ -1028,15 +1028,32 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheckResult = super.checkInputDataTypes()
+    if (defaultCheckResult.isFailure) {
+      return defaultCheckResult
+    }
+
+    trimStr match {
+      case None => TypeCheckResult.TypeCheckSuccess
+      case Some(trimChars) =>
+        val collationId = srcStr.dataType.asInstanceOf[StringType].collationId
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1052]: I think you already have this above as `final lazy val`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -1054,32 +1071,64 @@ trait String2TrimExpression extends Expression with ImplicitCastInputTypes {
...
+          |} else {
+          |  ${trimString.code}
+          |  if (${trimString.isNull}) {
+          |    ${ev.isNull} = true;
+          |  } else {
+          |    ${ev.value} =
+          |      ${srcString.value}.$trimMethod(${trimString.value}, $collationId);
+          |  }
+          |}""".stripMargin)
+      }
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1131]: I think this code is very similar to the one above, consider rewriting it a bit more neatly in a way that incorporates $collationId

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -1170,12 +1219,25 @@ case class StringTrim(srcStr: Expression, trimStr: Option[Expression] = None)
 
   override protected def direction: String = "BOTH"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trim()
+  override val trimMethod: String = "trim"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1222]: avoid unnecessary changes (I think this was below `override def doEval`)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -1278,12 +1340,25 @@ case class StringTrimLeft(srcStr: Expression, trimStr: Option[Expression] = None
 
   override protected def direction: String = "LEADING"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trimLeft()
+  override val trimMethod: String = "trimLeft"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1343]: ditto (avoid unnecessary changes)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 8]
```diff
@@ -1339,12 +1414,25 @@ case class StringTrimRight(srcStr: Expression, trimStr: Option[Expression] = Non
 
   override protected def direction: String = "TRAILING"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trimRight()
+  override val trimMethod: String = "trimRight"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1417]: ditto (avoid unnecessary changes)

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 9]
```diff
@@ -686,6 +743,111 @@ public UTF8String trimLeft(UTF8String trimString) {
...
+      return this;
+    }
+    if (trimByteIdx  >= numBytes) {
+      // Everything trimmed
+      return EMPTY_UTF8;
+    }
+    return copyUTF8String(trimByteIdx, numBytes - 1);
+  }
+
+  private UTF8String collatedTrimLeft(UTF8String trimString, int collationId) {
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:808]: @uros-db I see a number of PRs from different people adding similar functionality so just wondering if you can start a discussion to standardize the names of these kinds of methods because I've seen multiple variants

my vote would be for `collationAwareOp`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 10]
```diff
@@ -767,6 +947,139 @@ public UTF8String trimRight(UTF8String trimString) {
...
+  /**
+   * Trims characters of the given trim string from the end of this string.
+   * This variant of the method additionally applies provided collation to this string
+   * and trim characters before searching.
+   *
+   * @param trimString The trim characters string.
+   * @param collationId Id of the collation to use.
+   * @return this string with no occurrences of the trim characters at the end.
+   */
+  public UTF8String trimRight(UTF8String trimString, int collationId) {
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:959]: as far as I see this is the only public method exposing this trim api, so does that mean that we will always have to perform collation check in order to call collated trim, even if in codegen we know which collation we are working with?

this seems like a perf issue when you consider running this for millions of rows

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 11]
```diff
@@ -686,6 +743,111 @@ public UTF8String trimLeft(UTF8String trimString) {
...
+      return null;
+    }
+
+    // The searching byte position in the lowercase source string
+    int searchIdx = 0;
+    // The byte position of a first non-matching character in the lowercase source string
+    int trimByteIdx = 0;
+
+    // Convert trimString to lowercase so it can be searched properly
+    trimString = trimString.toLowerCase();
```
[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:778]: yes, in the context of string expressions, we use `.toLowerCase()` / `.toUpperCase()` only for _UTF8_BINARY_LCASE_ so we should consider specifying the base locale (**Locale.ROOT**) in those calls

[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:778]: not in the context of string expressions, here we use custom `UTF8String` functions:
```
public UTF8String toLowerCase()
public UTF8String toUpperCase()
```

note that in this context we only use this for UTF8_BINARY_LCASE, so no locale to be specified

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 9 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:808]: `collationAwareFunctionName` sounds good to me, we can discuss consolidating the naming for all functions

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 12]
```diff
@@ -628,6 +646,27 @@ public UTF8String trim(UTF8String trimString) {
...
+  /**
+   * Trims characters of the given trim string from both ends of this string.
+   * This variant of the method additionally applies provided collation to this string
+   * and trim characters before searching.
+   *
+   * @param trimString The trim characters string.
+   * @param collationId Id of the collation to use.
+   * @return this string with no occurrences of the characters from trim string.
+   */
+  public UTF8String trim(UTF8String trimString, int collationId) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:658]: Let's go with this naming: `collationAwareTrim`

ref: https://github.com/apache/spark/pull/45749#discussion_r1546421174

**Comment by uros-db** [comment]
heads up: we’ve done some major code restructuring in https://github.com/apache/spark/pull/45978, so please sync these changes before moving on

@davidm-db you’ll likely need to rewrite the code in this PR, so please follow the guidelines outlined in https://issues.apache.org/jira/browse/SPARK-47410

**Comment by mihailom-db** [comment]
Please add these expressions to CollationTypeCasts rules.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 13]
```diff
@@ -175,12 +175,19 @@ public Collation(
...
+      final UTF8String targetUTF8String,
+      final UTF8String patternUTF8String,
       final int collationId) {
-    String pattern = right.toString();
-    CharacterIterator target = new StringCharacterIterator(left.toString());
+    String pattern = patternUTF8String.toString();
+    CharacterIterator target = new StringCharacterIterator(targetUTF8String.toString());
     Collator collator = CollationFactory.fetchCollation(collationId).collator;
     return new StringSearch(pattern, target, (RuleBasedCollator) collator);
   }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:193]: did you pull the latest changes on master? we now have a different version of stringSearch

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:193]: https://github.com/apache/spark/blob/b4624bf4be28974e8df175670f89cf96858b7b81/common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java#L195

**Review by uros-db — COMMENTED** [review]
please change the PR title so we can remove it from the corresponding [ticket](https://issues.apache.org/jira/browse/SPARK-47409) in favor of your [new PR](https://github.com/apache/spark/pull/46206)
